test/e2e/README.md

```diff
@@ -0,0 +1,363 @@
+            ))
+        })
+    })
+})
+```
+
+Similarly to closure variables, `BeforeEach` blocks are typically defined at the top of the most relevant `Describe`,
+`Context` or `When` block.
+
+**:information_source: In the context of our framework, a new instance of `framework.Framework`, and therefore a new
+Kubernetes namespace, is created for each `It` block.** Consider this when writing tests that create API objects which
+take a long time to set up. Here is an [example test][optimized-test] optimized to run each of its specs in the same
+namespace.
+
+#### Fail within helpers
+
+Keeping the number of lines of code to a minimum within the main test body helps focusing on the actual logic. Resorting
+to helper functions is a great way to achieve this, and handling errors directly within those helpers is equally
+important.
+
+Generally speaking, if a function returns an error along with other values and we don't expect an error to occur within
+tests, that function can be wrapped inside a helper which only returns the values that are meant to be used in tests.
+Instead of leaving the error handling to `It` or `BeforeEach` blocks, the helper can call `framework.Failf` whenever an
+error occurs and immediately fail the current test.
+
+```go
+// okay-ish
+
+var _ = Describe("Error handling", func() {
+    It("does something and checks the output", func() {
+        something, err := mypkg.GetSomething()
+        Expect(err).ToNot(HaveOccured())
+
+        output, err := mypkg.HandleSomething(something)
+        Expect(err).ToNot(HaveOccured())
+        Expect(output).ToNot(BeNil())
+    })
+})
+```
+
+```go
+// better
+
+var _ = Describe("Error handling", func() {
+    It("does something and checks the output", func() {
+        Expect(doSomething()).ToNot(BeNil())
+    })
+})
+
+func doSomething() *mypkg.Output {
+    something, err := mypkg.GetSomething()
+    if err != nil {
+        framework.Failf("Failed to get something: %s", err)
+    }
+
+    output, err := mypkg.HandleSomething(something)
+    if err != nil {
+        framework.Failf("Failed to handle something: %s", err)
+    }
+
+    return output
+}
+```
+
+The benefits of this approach become more obvious when the complexity of test scenarios grows beyond the example above.
+
+
+[ginkgo-docs]: https://onsi.github.io/ginkgo/
+[ginkgo-issue222]: https://github.com/onsi/ginkgo/issues/222
+[ginkgo-struct]: https://onsi.github.io/ginkgo/#structuring-your-specs
+[optimized-test]: https://github.com/triggermesh/test-infra/blob/956c8ce257/test/e2e/sources/awscodecommit/main.go#L172-L188
+
+[k8s-e2e]: https://godoc.org/k8s.io/kubernetes/test/e2e
+[sigtesting-howto]: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/writing-good-e2e-tests.md
+
+[aws-go-session]: https://docs.aws.amazon.com/sdk-for-go/api/aws/session/
+[gh-client]: https://github.com/triggermesh/test-infra/blob/9f19ed28a9/test/e2e/framework/github/github.go#L39-L53
```

Comment:
should the links pointing to the old `test-infra` be updated to point to this repo?

Reply:
Eventually yes, but the resources linked here haven't yet be migrated, so I will need to do that _after_ migrating them.